### ISSUE TITLE
Trouble detecting correct OpenGL version

### ISSUE THREAD

**Issue by DeadNews** [mpv-player/mpv#9532]
### Important Information

- mpv version: 0.34.0
- Linux Distribution and Version: Manjaro Stable
- Source of the mpv binary: Manjaro repo / git
- If known which version of mpv introduced the problem: 0.34.0
- Window Manager and version: KWin 5.23.3
- GPU driver and version: NVIDIA 495.44

### Reproduction steps
```sh
mpv --gpu-debug=yes --v video.file
```

### Expected behavior
Works with:
```ini
glsl-shaders='~~/shaders/compute/nnedi3-nns16-win8x4.hook'
```

### Actual behavior
Blue screen
```
[vo/gpu/opengl] 0(58) : error C7532: global function intBitsToFloat requires "#version 330" or later
```
https://en.wikipedia.org/wiki/OpenGL_Shading_Language#Versions

### Log file
With v0.33.1 (works fine):
```
[vo/gpu/opengl] Initializing GPU context 'x11egl'
[vo/gpu/x11] X11 opening display: :0
[vo/gpu/x11] Display 0 (DP-0): [0, 0, 3840, 2160] @ 59.996625 FPS
[vo/gpu/x11] Current display FPS: 59.996625
[vo/gpu/opengl] EGL_VERSION=1.5
[vo/gpu/opengl] EGL_VENDOR=NVIDIA
[vo/gpu/opengl] EGL_CLIENT_APIS=OpenGL_ES OpenGL
[vo/gpu/opengl] Trying to create Desktop OpenGL context.
[vo/gpu/opengl] Choosing visual EGL config 0x28, visual ID 0x2b
[vo/gpu/opengl] GL_VERSION='4.4.0 NVIDIA 495.44'
[vo/gpu/opengl] Detected desktop OpenGL 4.4.
[vo/gpu/opengl] GL_VENDOR='NVIDIA Corporation'
[vo/gpu/opengl] GL_RENDERER='NVIDIA GeForce GTX 1050/PCIe/SSE2'
[vo/gpu/opengl] GL_SHADING_LANGUAGE_VERSION='4.40 NVIDIA via Cg compiler'
[vo/gpu] Testing FBO format rgba16f
[vo/gpu] Using FBO format rgba16f.
```

With v0.34.0 from Manjaro repo / git upstream:
```
vo/gpu/opengl] Initializing GPU context 'x11egl'
[vo/gpu/x11] X11 opening display: :0
[vo/gpu/x11] Display 0 (DP-0): [0, 0, 3840, 2160] @ 59.996625 FPS
[vo/gpu/x11] Current display FPS: 59.996625
[vo/gpu/opengl] EGL_VERSION=1.5
[vo/gpu/opengl] EGL_VENDOR=NVIDIA
[vo/gpu/opengl] EGL_CLIENT_APIS=OpenGL_ES OpenGL
[vo/gpu/opengl] Trying to create Desktop OpenGL context.
[vo/gpu/opengl] Choosing visual EGL config 0x28, visual ID 0x2b
[vo/gpu/opengl] GL_VERSION='3.2.0 NVIDIA 495.44'
[vo/gpu/opengl] Detected desktop OpenGL 3.2.
[vo/gpu/opengl] GL_VENDOR='NVIDIA Corporation'
[vo/gpu/opengl] GL_RENDERER='NVIDIA GeForce GTX 1050/PCIe/SSE2'
[vo/gpu/opengl] GL_SHADING_LANGUAGE_VERSION='1.50 NVIDIA via Cg compiler'
[vo/gpu/opengl] Loaded extension GL_ARB_timer_query.
[vo/gpu/opengl] Loaded extension GL_ARB_get_program_binary.
[vo/gpu/opengl] Loaded extension GL_ARB_buffer_storage.
[vo/gpu/opengl] Loaded extension GL_ARB_shader_image_load_store.
[vo/gpu/opengl] Loaded extension GL_ARB_shader_storage_buffer_object.
[vo/gpu/opengl] Loaded extension GL_ARB_compute_shader.
[vo/gpu/opengl] Loaded extension GL_ARB_arrays_of_arrays.
[vo/gpu/opengl] Loaded extension GL_ARB_debug_output.
[vo/gpu] Testing FBO format rgba16f
[vo/gpu] Using FBO format rgba16f.
``` 
Glxinfo output:
```
-> glxinfo | grep -i opengl
OpenGL vendor string: NVIDIA Corporation
OpenGL renderer string: NVIDIA GeForce GTX 1050/PCIe/SSE2
OpenGL core profile version string: 4.6.0 NVIDIA 495.44
OpenGL core profile shading language version string: 4.60 NVIDIA
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 4.6.0 NVIDIA 495.44
OpenGL shading language version string: 4.60 NVIDIA
OpenGL context flags: (none)
OpenGL profile mask: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.2 NVIDIA 495.44
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
OpenGL ES profile extensions:
```

**Comment by Dudemanguy** [comment]
Hmm, this is related to the refactor of the EGL code. It was noted that NVIDIA could return funny versions. I hope we don't have to add in workarounds for them again.

**Comment by sfan5** [comment]
I pointed this out in https://github.com/mpv-player/mpv/pull/9268#issuecomment-940228290 and the conclusion was that it's not an issue for _mpv_ itself.

> Glxinfo output:

FYI mpv does not use glx anymore so this output is a hint but not directly relevant.

**Comment by evelikov** [comment]
From a quick look - this looks like a pre-existing bug in `mpv`.

In particular - the driver reports GLSL `1.50` as seen in `GL_SHADING_LANGUAGE_VERSION` yet we try to use `330` in the shader. Let me see how easy it is to fix...

**Comment by sfan5** [comment]
> Alternatively we can re-introduce the `440` at the top of the list...

Wouldn't that have the same issue with features added 3.2->3.3 or 4.4->4.6?
`320` is currently at the top of the list and this results in exactly a 3.2.0 context on my system.

**Comment by philipl** [comment]
This has additional problems on nvidia, regardless of shaders. The low OpenGL version means that no compute shaders are detected and then HDR peak detection is disabled. We do actually want a higher context version for compute shaders, so we should be trying to detect 440 (or whatever the actual right version is for that).

**Comment by philipl** [comment]
For the record, the minimum version to get compute shaders is 420

**Comment by CounterPillow** [comment]
> For the record, the minimum version to get compute shaders is 420

dank

But also, you shouldn't assume 420 is required, check for [`GL_ARB_compute_shader`](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_compute_shader.txt). E.g. panfrost implements compute shaders, but not geometry shaders, so has no OpenGL > 3.1 advertised but has compute shaders advertised as an extension.

**Comment by sfan5** [comment]
It looks like mpv does that actually...
https://github.com/mpv-player/mpv/blob/fc94c8c365ebeb038af6052bf4ea0506c1220559/video/out/opengl/common.c#L362-L367
https://github.com/mpv-player/mpv/blob/fc94c8c365ebeb038af6052bf4ea0506c1220559/video/out/opengl/common.c#L656-L658

...however I can confirm that it doesn't detect them correctly on my NVIDIA system:
`[vo/gpu] Disabling HDR peak computation (one or more of the following is not supported: compute shaders=0, SSBO=1).`

**Comment by philipl** [comment]
Yeah. At 420, compute shaders load as an extension and work. At 430 or higher it loads as core functionality and also works. Any lower version and it doesn't load. And on ES it doesn't load either, even though we get a 3.2 context.

At a guess, I'd assume nvidia has interpreted the spec in some way that says they shouldn't be exposed at lower versions?

**Comment by philipl** [comment]
The extension spec explicitly says "OpenGL 4.2 is required." And so nvidia have enforced that.

**Comment by evelikov** [comment]
I think that the biggest issue here is that mpv internal details have leaked into the user-supported shaders. In particular, any robust shader should not assume a `#version` but require/enable respective extensions.

Thad said, it seems like the ship has sailed and people rely on that behaviour - sigh.

**Comment by sfan5** [comment]
fixed by 0e76372e86d92982e17219fda6597f92ca67e7e8

**Comment by DeadNews** [comment]
Thanks, it works.
